crates/script/src/sequence.rs

```diff
@@ -26,6 +23,86 @@ use std::{
 };
 use yansi::Paint;
 
+/// Returns the commit hash of the project if it exists
+pub fn get_commit_hash(root: &Path) -> Option<String> {
+    Git::new(root).commit_hash(true, "HEAD").ok()
+}
+
+pub enum ScriptSequenceKind {
+    Single(ScriptSequence),
+    Multi(MultiChainSequence),
+}
+
+impl ScriptSequenceKind {
+    pub fn save(&mut self, silent: bool, save_ts: bool) -> Result<()> {
+        match self {
+            ScriptSequenceKind::Single(sequence) => sequence.save(silent, save_ts),
+            ScriptSequenceKind::Multi(sequence) => sequence.save(silent, save_ts),
+        }
+    }
+
+    pub fn sequences(&self) -> impl Iterator<Item = &ScriptSequence> {
+        match self {
+            ScriptSequenceKind::Single(sequence) => std::slice::from_ref(sequence).iter(),
+            ScriptSequenceKind::Multi(sequence) => sequence.deployments.iter(),
+        }
+    }
+
+    pub fn sequeneces_mut(&mut self) -> impl Iterator<Item = &mut ScriptSequence> {
+        match self {
+            ScriptSequenceKind::Single(sequence) => std::slice::from_mut(sequence).iter_mut(),
+            ScriptSequenceKind::Multi(sequence) => sequence.deployments.iter_mut(),
+        }
+    }
+
+    pub fn sequences_len(&self) -> usize {
+        match self {
+            ScriptSequenceKind::Single(_) => 1,
+            ScriptSequenceKind::Multi(sequence) => sequence.deployments.len(),
+        }
+    }
+
+    pub fn get_sequence_mut(&mut self, index: usize) -> Option<&mut ScriptSequence> {
```

Comment:
I think all these methods are redundant if we can just use `fn sequences() -> &[ScriptSequence]`

Reply:
The need for `get_sequence_mut` and `sequences_len` is from the `BundledState::broadcast` where we need mutable access to separate sequences to populate pending txes and receipts, but we also sometimes need mutable access to the parent `ScriptSequenceKind` to do checkpoint save of all sequences to the file. so not sure what's the best approach to reduce redundancy here

Reply:
right I mean this
```rust
    pub fn sequences(&self) -> &[ScriptSequence] {
        match self {
            ScriptSequenceKind::Single(sequence) => std::slice::from_ref(sequence),
            ScriptSequenceKind::Multi(sequence) => &sequence.deployments,
        }
    }

    pub fn sequences_mut(&mut self) -> &mut [ScriptSequence] {
        match self {
            ScriptSequenceKind::Single(sequence) => std::slice::from_mut(sequence),
            ScriptSequenceKind::Multi(sequence) => sequence.deployments,
        }
    }
```

Then all the other functions (`get_sequence_mut`, `sequences_len`) can just call the corresponding methods on the slices returned by these methods

Reply:
ah got it, updated in 5bd705d3ca951b6b0828fa7e4e74ed58c90b9814